### ISSUE TITLE
Don't abuse Singleton (sharedInstance) implementation patter

### ISSUE THREAD

**Issue by kostiakoval** [hyperoslo/iOS-playbook#139]
Singleton is abused patter and often you can achieve similar result by using dependency injection. 
Singleton Abuse

Example: - `[MYState  sharedState]`

Problem wit Singleton:
- Its globally available. It a global variables in the nicer class wrapper
- It hard to unit test. The methods that access global singleton need to be mocked
- provides predictable result. 
  One component can save some data to global singleton and other component wouldn't know that.

```swift
func getResult()-> Int {
  MyState.sharedState().result
}
```
For this example i need to mock `MyState.sharedState`  
**Solve problem** 
Instantiate Singleton in the init or other method and  pass it to the method that does the work.

```swift 
func getResult(state: MyState)-> Int {
  state.result
}
```
In that cases getResult doesn't know that MyState is an Singleton and in unit test you can make state object (not Singleton) and pass it.

 
**Resources :**
http://www.objc.io/issue-13/singletons.html
https://www.youtube.com/watch?v=-FRm3VPhseI&index=2&list=PL693EFD059797C21E
http://stackoverflow.com/questions/137975/what-is-so-bad-about-singletons

**Comment by 3lvis** [comment]
I agree, you can say that 99.99% of the cases using a Singleton is the wrong answer.

As I see it, Singleton is **most** of the time, a shortcut for understanding the application workflow.

And you shouldn't take that shortcut, it will bite you hard afterwards.
